### DISCUSSION TITLE
Custom Components

### DISCUSSION THREAD

**Discussion by nkawerau** [oemof/tespy#313] [Q&A]
I am currently trying to implement a new custom component in Tespy for an alkaline electrolysis. The basic steps are clear to me from the documentation. Nevertheless, I am having trouble setting up the method for the partial derivatives of the system equations for the Newton–Raphson solver correctly. I think the following section in the documentation tries to explain this: “For the derivative, the partial derivatives to all variables of the network are required. This means, that you have to calculate the partial derivatives to mass flow, pressure, enthalpy and all fluids in the fluid vector on each connection affecting the equation defined before”. However, it’s still not clear to me what the structure of the derivatives should look like. Please explain the differences between derivatives for component variables and for component equations. Some  methods “_deriv“ write directly to self.jacobian others return a numpy array called deriv.

**Comment by fwitte** [comment]
Hi @nkawerau,

thank you for reaching out, the docs are kind of outdated here and should be improvemed! 

> However, it’s still not clear to me what the structure of the derivatives should look like. Please explain the differences between derivatives for component variables and for component equations. Some methods “_deriv“ write directly to self.jacobian others return a numpy array called deriv.

There are two different types of derivative methods:

1. **Constant** derivatives for mandatory equations (independet of the value of component parameters, e.g. mass balance)
2. **Not constant** derivatives for mandatory equations or parameter related equations (such equations that are only applied, if the corresponding parameter has been specified by the user)

For example,

1. The mass balance of a single inlet and outlet component (turbine, compressor, ...). Here the data are set up and written in to the Jacobian prior to solving the network, because their value will never change https://github.com/oemof/tespy/blob/faf5a8c87480d4590ea0efa3d1b65ea2e58bfd42/src/tespy/components/component.py#L1035-L1053
2. The energy balance equation for a HeatExchanger (energy transferred from hot side equals energy transferred to cold side in magnitude) https://github.com/oemof/tespy/blob/faf5a8c87480d4590ea0efa3d1b65ea2e58bfd42/src/tespy/components/heat_exchangers/base.py#L302-L320, or the energy balance equation of a turbomachine, where the power of the component is specified https://github.com/oemof/tespy/blob/faf5a8c87480d4590ea0efa3d1b65ea2e58bfd42/src/tespy/components/turbomachinery/base.py#L140-L154 Here the data need an update for every Newton iteration, therefore the data are updated directly in the Jacobian.

Hope it helps :). How would the equation(s) you want to implement look like?
